### ISSUE TITLE
Disable relaxed math testing for CL1.2 (or loosen up the ULP requirements)

### ISSUE THREAD

**Issue by aarongreig** [KhronosGroup/OpenCL-CTS#703]
The OpenCL 2.0 spec changed the definition of `-cl-fast-relaxed-math` to include ULP requirements for certain operations. I initially opened #702 to avoid testing the most stringent of these new ULP requirements with CL1.2 implementations, but ideally a solution should be found that can easily scale to all of the operations listed in table 38 under [numerical compliance](https://www.khronos.org/registry/OpenCL/specs/2.2/html/OpenCL_C.html#relative-error-as-ulps). The simplest solution would be to change the 1.2 test CSVs to pass `-r` to the bruteforce executable, disabling relaxed altogether. I'm still looking to see if there might be a way to retain some relaxed math testing with 1.2, sans the new ULP requirements.

**Comment by alycm** [comment]
Does anyone have any suggestions on how approach this?  It affects 1.2 implementations attempting to use master branch to test.

The spreadsheet based approach is the simplest, but doesn't seem clean to me as then running the bruteforce binary manually may still fail if that option is not passed.

It seems like the test framework could be hard-coded to "know" to skip relaxed variants for OpenCL 1.x, but that also doesn't seem very clean because otherwise the version information is in the test description.

Another option could be that in addition to `RELAXED_ON` and `RELAXED_OFF` there is also a `RELAXED_ON_GT_CL12` option that test configurations could use where appropriate?

**Comment by kpet** [comment]
>Another option could be that in addition to RELAXED_ON and RELAXED_OFF there is also a RELAXED_ON_GT_CL12 option that test configurations could use where appropriate?

Having given the code a quick look, I think that's a more attractive approach. We should do our best to keep all the information in the tables to make things easier to cross-check with the spec.

As a pre-requisite, I think we should change how `gTestFastRelaxed`  is used. The global variable is set and read from a number of places which makes it hard to reason about the overall behaviour. I suggest that we pass the requirement to test in relaxed mode as a parameter to the test function rather than have the sequencing code do horrible things like https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/math_brute_force/main.cpp#L217.

**Comment by FranklandJack** [comment]
I have opened a PR to address the issues related to the global variable `gTestFastRelaxed`: https://github.com/KhronosGroup/OpenCL-CTS/pull/807.

**Comment by FranklandJack** [comment]
Would the cleanest approach to solving the problem of the different precision requirements between 2.0 and earlier versions not just be to skip the relaxed testing completely for 1.0, 1.1 and 1.2? If we could skip the tests from [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/math_brute_force/main.cpp#L176) based on the device version, then that would be consistent with https://www.khronos.org/registry/OpenCL/specs/opencl-1.2.pdf which says that for `-cl-fast-relaxed-math` 

> This allows optimizations for floating-point arithmetic that may violate the IEEE 754 standard and the OpenCL numerical compliance requirements defined in in section 7.4 for single precision and double precision floating-point, and edge case behavior in section 7.5. 

which implies we can't reliably test the precision for relaxed math anyway since its not defined.

**Comment by alycm** [comment]
A flag like `RELAXED_ON_GT_CL12` seems clearer to me, because then the test source makes it clear when relaxed math is tested.   Burying the check deep inside the `doTest` function does not seem clear.

So yes, the relaxed tests would be skipped for devices <= 1.2, but due to an explicit test parameter, not due to a hidden check.

**Comment by FranklandJack** [comment]
> A flag like `RELAXED_ON_GT_CL12` seems clearer to me, because then the test source makes it clear when relaxed math is tested. Burying the check deep inside the `doTest` function does not seem clear.
> 
> So yes, the relaxed tests would be skipped for devices <= 1.2, but due to an explicit test parameter, not due to a hidden check.

I agree that it is clearer but we already have a `relaxed` field in [the struct](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/math_brute_force/FunctionList.h#L87) that is checked [here](https://github.com/KhronosGroup/OpenCL-CTS/blob/master/test_conformance/math_brute_force/main.cpp#L178) to determine whether this test will run or not when the relaxed tests are run. Since it only makes sense to run relaxed testing for versions later than 1.2, couldn't we just check the device version when we check if this option is set?

**Comment by alycm** [comment]
> Since it only makes sense to run relaxed testing for versions later than 1.2, couldn't we just check the device version when we check if this option is set?

I'm not sure exactly how it would look in the code, but are we proposing the same thing?  Of course if a `RELAXED_ON_GT_CL12` flag is used then the version needs checked somewhere.

My comments about burying the check in that `doTest` function is if the test configuration says `RELAXED_ON`, and someone is testing a 1.2 device then it seems a bit arbitrary that this is ignored.  Encoding the "this is version dependent" information in the configuration seems more readable than encoding it in the code.

So I would imagine that the check looks something like:

```c
if ((relaxed == RELAXED) || (relaxed == RELAXED_ON_GT_CL12 && version > 120)) {
  // Test relaxed math
}
```

I'm not especially familar with this code though, so I don't know the version can be easily checked there nor whether recent framework changes provide something to help.

**Comment by alycm** [comment]
Internal discussion has lead me to realize that all instances of `RELAXED_ON` would become `RELAXED_ON_GT_CL12` in my proposal.

Thus my idea just amounts to a rename, which may or may not be worth doing.

**Comment by FranklandJack** [comment]
I've opened#819 which addresses the above comments.

**Comment by rjodinchr** [comment]
Fixed by PR #819 (commit `96ea9d60`), which skips reduced precision math testing for OpenCL devices < 2.0 because OpenCL specifications prior to 2.0 did not define specific ULP error limits for `-cl-fast-relaxed-math` operations.
